### PR TITLE
support internvl vit

### PR DESCRIPTION
**Pull request by Jzz24**

### CODE REVIEW

File: model_zoo/intern_vit/huggingface/ConvertInternVLWeightToOPmx.py
Comment (Alcanderian) [inline model_zoo/intern_vit/huggingface/ConvertInternVLWeightToOPmx.py:1]: 这个文件已经冗余，可以删除

File: torch_function/TensorParallelRMSNorm.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                 input_is_parallel_i = input_is_parallel,
+                 )
+        return Y.setTypeAs(X)
+
+
+    @staticmethod
+    def forward(
+        self, X: torch.Tensor, weight: torch.Tensor, proc_group: dist.ProcessGroup,
+        axis: int = -1, eps: float = 1e-5,
+        embed_dim: int = 128, scale: float = 1.0,
```
Comment (Alcanderian) [inline torch_function/TensorParallelRMSNorm.py:25]: embed dim不需要

计算逻辑
tp_variance = x_parallel.pow(2).mean(axis, keepdim=True)
if proc_group is not None and torch.distributed.get_world_size(proc_group) > 1:
       dist.all_reduce(tp_variance, op=dist.ReduceOp.SUM, async_op=False)
variance = tp_variance / (world_size * scale)

File: model_zoo/intern_vit/huggingface/ConvertInternVLPadWeightToOPmx.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--input_dir",
+        help="Location of HF weights, which contains tokenizer.model and model folders",
+    )
+    parser.add_argument(
+        "--output_dir",
+        help="Location to write PMX model",
+    )
+    args = parser.parse_args()
```
Comment (Alcanderian) [inline model_zoo/intern_vit/huggingface/ConvertInternVLPadWeightToOPmx.py:167]: 添加pad_to_head参数，默认1

padded_num_heads = (num_heads + pad_to_head - 1) // pad_to_head * pad_to_head

File: model_zoo/intern_vit/huggingface/ConvertInternVLPadWeightToOPmx.py
Comment (Alcanderian) [inline model_zoo/intern_vit/huggingface/ConvertInternVLPadWeightToOPmx.py:1]: 重命名

File: model_zoo/intern_vit/modeling/Params.py [hunk 3]
```diff
@@ -15,3 +15,7 @@ class ViTParams:
     image_size: int = 224
     patch_size: int = 32
     projection_dim: int = 512
+
+    llm_hidden_dim: int = 512
+    downsample_ratio: int = 2
+    qk_norm: bool = True
```
Comment (Alcanderian) [inline model_zoo/intern_vit/modeling/Params.py:21]: 根据pad添加 qk_norm_scale，padded_num_heads, padded_num_kv_heads，head_dim

File: model_zoo/intern_vit/huggingface/ConvertInternVLWeightToOPmx.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        "vision_projection.w1.bias": hf_model_state_dict["mlp1.1.bias"],
+        "vision_projection.w2.weight": hf_model_state_dict["mlp1.3.weight"],
+        "vision_projection.w2.bias": hf_model_state_dict["mlp1.3.bias"],
+    })
+
+    if "visual_projection.weight" in hf_model_state_dict.keys():
+        state_dict.update({"vision_projection.weight": hf_model_state_dict["visual_projection.weight"]})
+
+    torch.save(state_dict, os.path.join(model_path, "model.pth"))
+
```
Comment (Alcanderian) [inline model_zoo/intern_vit/huggingface/ConvertInternVLWeightToOPmx.py:133]: 导出带有padded_num_heads和padded_num_kv_heads，还有qk_norm_scale, head_dim的json

File: model_zoo/intern_vit/modeling/static_batching/Model.py [hunk 5]
```diff
@@ -53,14 +105,16 @@ def __init__(
         self.num_local_heads = args.num_heads // world_size
         self.num_local_kv_heads = self.num_kv_heads // world_size
         self.num_local_kv_repeats = self.num_local_heads // self.num_local_kv_heads
-        self.head_dim = args.hidden_dim // args.num_heads
+        #self.head_dim = args.hidden_dim // args.num_heads
+        self.head_dim = 128 # tmp fix for pad num_head=25
```
Comment (Alcanderian) [inline model_zoo/intern_vit/modeling/static_batching/Model.py:109]: 读取 json的head_dim, 其他位置一样

File: model_zoo/intern_vit/modeling/static_batching/Model.py [hunk 6]
```diff
@@ -14,11 +14,63 @@
...
 
 TensorDumper = ModelUtils.__TensorDumper__()
 
 
+class TensorParallelRMSNorm(nn.Module):
+    def __init__(self, proc_group: dist.ProcessGroup, dim: int, eps: float = 1e-5,
+                 scale: float = 1.0, input_is_parallel: bool = False) -> None:
+        super().__init__()
+        self.eps = eps
+        self.embed_dim = dim
```
Comment (Alcanderian) [inline model_zoo/intern_vit/modeling/static_batching/Model.py:28]: 不需要embed_dim

**Review by Alcanderian — CHANGES_REQUESTED** [review]

File: model_zoo/intern_vit/huggingface/ConvertInternVLPadWeightToOPmx.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        "vision_projection.w1.bias": hf_model_state_dict["mlp1.1.bias"],
+        "vision_projection.w2.weight": hf_model_state_dict["mlp1.3.weight"],
+        "vision_projection.w2.bias": hf_model_state_dict["mlp1.3.bias"],
+    })
+
+    if "visual_projection.weight" in hf_model_state_dict.keys():
+        state_dict.update({"vision_projection.weight": hf_model_state_dict["visual_projection.weight"]})
+
+    torch.save(state_dict, os.path.join(model_path, "model.pth"))
+
```
Comment (Alcanderian) [inline model_zoo/intern_vit/huggingface/ConvertInternVLPadWeightToOPmx.py:155]: 导出带有padded_num_heads和padded_num_kv_heads，还有qk_norm_scale, head_dim的json

命名为opmx_params.json

同时要照顾原始的ConvertWeightToOPMX导出json
可以开一个新的intern_vl文件夹，里面带有vit的子文件夹
结构树
intern_vit
  ---- huggingface
          ---- convert_weight
intern_vl
  ---- huggingface
         ---- vit
               ---- convert_weight

File: model_zoo/intern_vit/huggingface/ConvertInternVLWeightToOPmx.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    # TO DO: GQA / MQA, only test on llama
+    num_heads = pmx_params_dict['num_heads']
+    num_kv_heads = pmx_params_dict['num_kv_heads']
+    dims_per_head = pmx_params_dict['hidden_dim'] // num_heads
+    key_value_dim = dims_per_head * num_kv_heads
+
+    # process pad params
+    pmx_params_dict['padded_num_heads'] = (num_heads + pad_to_head - 1) // pad_to_head * pad_to_head
+    pmx_params_dict['qk_norm_scale'] = num_heads / pmx_params_dict['padded_num_heads']
```
Comment (Alcanderian) [inline model_zoo/intern_vit/huggingface/ConvertInternVLWeightToOPmx.py:77]: 除反了？

**Review by Alcanderian — CHANGES_REQUESTED** [review]

File: model_zoo/intern_vit/modeling/static_batching/Model.py [hunk 9]
```diff
@@ -242,15 +292,15 @@ def forward(self, pixel_values: torch.Tensor, attn_mask: Optional[torch.Tensor])
...
-        output = (norm + h)[:, 0, :] # get cls token
+        # output = (norm + h)[:, 0, :] # get cls token
         # TensorDumper.dump(pooled_output, "pooled_output")
-        # output = self.post_layernorm(pooled_output)
-        # TensorDumper.dump(output, "post_layernorm_out")
+
         if self.with_proj_head:
-            output = self.vision_projection(output)
+            output = self.vision_projection((norm+h)[:, 1:, :])
             # TensorDumper.dump(output, "vision_proj_out")
```
Comment (Jzz24) [inline model_zoo/intern_vit/modeling/static_batching/Model.py:304]: else

**Review by Alcanderian — APPROVED** [review]
